src/rpc/masternode.cpp

```diff
@@ -296,31 +328,42 @@ UniValue masternode_winners(const JSONRPCRequest& request)
     if (request.fHelp || request.params.size() > 3)
         masternode_winners_help();
 
-    int nHeight;
+    const CBlockIndex* pindexTip{nullptr};
     {
         LOCK(cs_main);
-        CBlockIndex* pindex = chainActive.Tip();
-        if (!pindex) return NullUniValue;
-
-        nHeight = pindex->nHeight;
+        pindexTip = chainActive.Tip();
+        if (!pindexTip) return NullUniValue;
     }
 
-    int nLast = 10;
+    int nCount = 10;
     std::string strFilter = "";
 
     if (!request.params[1].isNull()) {
-        nLast = atoi(request.params[1].get_str());
+        nCount = atoi(request.params[1].get_str());
     }
 
     if (!request.params[2].isNull()) {
         strFilter = request.params[2].get_str();
     }
 
     UniValue obj(UniValue::VOBJ);
-    auto mapPayments = GetRequiredPaymentsStrings(nHeight - nLast, nHeight + 20);
-    for (const auto &p : mapPayments) {
-        if (strFilter != "" && p.second.find(strFilter) == std::string::npos) continue;
-        obj.pushKV(strprintf("%d", p.first), p.second);
+
+    int nChainTipHeight = pindexTip->nHeight;
+    int nStartHeight = std::max(nChainTipHeight - nCount, 1);
+
+    for(int h = nStartHeight; h <= nChainTipHeight; h++) {
+        auto payee = deterministicMNManager->GetListForBlock(pindexTip->GetAncestor(h - 1)).GetMNPayee();
+        std::string strPayments = GetRequiredPaymentsString(h, payee);
+        if (strFilter != "" && strPayments.find(strFilter) == std::string::npos) continue;
+        obj.pushKV(strprintf("%d", h), strPayments);
+    }
+
+    auto projection = deterministicMNManager->GetListForBlock(pindexTip).GetProjectedMNPayees(20);
+    for (size_t i = 0; i < projection.size(); i++) {
+        int h = nChainTipHeight + 1 + i;
+        std::string strPayments = GetRequiredPaymentsString(h, projection[i]);
+        if (strFilter != "" && strPayments.find(strFilter) == std::string::npos) continue;
+        obj.pushKV(strprintf("%d", h), strPayments);
     }
```

Comment:
Can you explain what you're doing in these changes?

Reply:
It's basically a simplified version of https://github.com/dashpay/dash/pull/3861/files#diff-d99314985bd89b12d74bb2e64afd73ebbb5483249b862ac5a07e5edba98ce8edL251-L282 with all the unnecessary parts dropped (we know `nEndHeight` is always `nChainTipHeight + 20`).